### PR TITLE
feat: Implement environment update functionality in launchdevly ui

### PR DESCRIPTION
**Pull request by cdelst**
https://www.loom.com/share/547246f3f0384ff69260bd744e9dd392

image

### CODE REVIEW

File: internal/dev_server/api/server.go [hunk 1]
```diff
@@ -8,14 +8,16 @@ import (
 )
 
 //go:generate go run github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen -config oapi-codegen-cfg.yaml api.yaml
-type Server struct {
+var _ StrictServerInterface = server{}
+
+type server struct {
```
Comment (cdelst) [inline internal/dev_server/api/server.go:13]: this should be private

File: internal/dev_server/api/server.go [hunk 2]
```diff
@@ -8,14 +8,16 @@ import (
...
 }
 
-func NewStrictServer() Server {
-	return Server{}
+func NewStrictServer() StrictServerInterface {
```
Comment (cdelst) [inline internal/dev_server/api/server.go:16]: we want to return the interface not the struct

File: internal/dev_server/api/server.go [hunk 3]
```diff
@@ -8,14 +8,16 @@ import (
 )
 
 //go:generate go run github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen -config oapi-codegen-cfg.yaml api.yaml
-type Server struct {
+var _ StrictServerInterface = server{}
```
Comment (cdelst) [inline internal/dev_server/api/server.go:11]: compile check that's clear

File: internal/dev_server/api/server.go [hunk 4]
```diff
@@ -299,3 +301,29 @@ func (s Server) PutDevProjectsProjectKeyOverridesFlagKey(ctx context.Context, re
...
+	store := model.StoreFromContext(ctx)
+	project, err := store.GetDevProject(ctx, request.ProjectKey)
+	if err != nil {
+		return nil, err
+	}
+	if project == nil {
+		return GetProjectsEnvironments404JSONResponse{}, nil
+	}
+
+	environments, err := project.Environments(ctx)
```
Comment (cdelst) [inline internal/dev_server/api/server.go:315]: I need to refactor this off the project rep still.  On my list

File: internal/dev_server/db/sqlite.go [hunk 5]
```diff
@@ -83,9 +83,9 @@ func (s Sqlite) UpdateProject(ctx context.Context, project model.Project) (bool,
 	}()
 	result, err := tx.ExecContext(ctx, `
 		UPDATE projects
-		SET flag_state = ?, last_sync_time = ?, context=?
+		SET flag_state = ?, last_sync_time = ?, context=?, source_environment_key=?
```
Comment (cdelst) [inline internal/dev_server/db/sqlite.go:86]: was this intentionally left out? I saw some tests that might suggest so.

File: internal/dev_server/dev_server.go [hunk 6]
```diff
@@ -66,7 +67,7 @@ func (c LDClient) RunServer(ctx context.Context, serverParams ServerParams) {
 	handler = handlers.RecoveryHandler(handlers.PrintRecoveryStack(true))(handler)
 
 	addr := fmt.Sprintf("0.0.0.0:%s", serverParams.Port)
-	log.Printf("Server running on %s", addr)
+	log.Printf("server running on %s", addr)
```
Comment (cdelst) [inline internal/dev_server/dev_server.go:70]:
```suggestion
	log.Printf("Server running on %s", addr)
```

File: internal/dev_server/dev_server.go [hunk 7]
```diff
@@ -77,6 +78,7 @@ func (c LDClient) RunServer(ctx context.Context, serverParams ServerParams) {
 
 func getDBPath() string {
 	dbFilePath, err := xdg.StateFile("ldcli/dev_server.db")
+	log.Printf("Using database at %s", dbFilePath)
```
Comment (cdelst) [inline internal/dev_server/dev_server.go:81]: useful for locally connecting to the db

File: internal/dev_server/ui/src/App.tsx [hunk 8]
```diff
@@ -15,20 +15,30 @@ import {
 import { Icon } from '@launchpad-ui/icons';
 import { FlagVariation } from './api.ts';
 import { apiRoute, sortFlags } from './util.ts';
+import { ProjectEditor } from './ProjectEditor';
+
+// Add this interface
```
Comment (cdelst) [inline internal/dev_server/ui/src/App.tsx:20]:
```suggestion
```

File: internal/dev_server/adapters/api.go [hunk 9]
```diff
@@ -87,6 +98,14 @@ func (a apiClientApi) getFlags(ctx context.Context, projectKey string, href *str
 	return flags, nil
 }
 
+func (a apiClientApi) getEnvironments(ctx context.Context, projectKey string) ([]ldapi.Environment, error) {
```
Comment (mike-zorn) [inline internal/dev_server/adapters/api.go:101]: Probably should page through. You can use the `getFlags` as an example.

Comment (cdelst) [reply internal/dev_server/adapters/api.go:101]: yep, good call.

Comment (cdelst) [reply internal/dev_server/adapters/api.go]: added.  Also made the paging logic generic.

File: internal/dev_server/ui/src/App.tsx [hunk 10]
```diff
@@ -145,4 +226,15 @@ function App() {
   );
 }
 
+// Add this function at the end of the file
```
Comment (mike-zorn) [inline internal/dev_server/ui/src/App.tsx:229]: what?

**Review by mike-zorn — APPROVED** [review]

File: internal/dev_server/ui/src/App.tsx (hunk 10 above)
Comment (cdelst) [reply internal/dev_server/ui/src/App.tsx:229]: lol. Removing

**Pull request merged**